pkg/controller/keycloakclient/keycloakclient_reconciler.go

```diff
@@ -43,9 +43,43 @@ func (i *KeycloakClientReconciler) Reconcile(state *common.ClientState, cr *kc.K
 		desired.AddAction(i.getUpdatedClientSecretState(state, cr))
 	}
 
+	i.ReconcileRoles(state, cr, &desired)
+
 	return desired
 }
 
+func (i *KeycloakClientReconciler) ReconcileRoles(state *common.ClientState, cr *kc.KeycloakClient, desired *common.DesiredClusterState) {
+	rolesNew := roleDifference(cr.Spec.Roles, state.Roles)
+	for _, role := range rolesNew {
+		desired.AddAction(i.getCreatedClientRoleState(state, cr, &role))
+	}
+
+	rolesDeleted := roleDifference(state.Roles, cr.Spec.Roles)
+	for _, role := range rolesDeleted {
+		desired.AddAction(i.getDeletedClientRoleState(state, cr, &role))
+	}
+
+	rolesUpdated := roleDifference(roleDifference(cr.Spec.Roles, rolesNew), rolesDeleted)
+	for _, role := range rolesUpdated {
+		desired.AddAction(i.getUpdatedClientRoleState(state, cr, &role))
+	}
+}
+
+func roleDifference(a []kc.RoleRepresentation, b []kc.RoleRepresentation) (d []kc.RoleRepresentation) {
+	remove := make(map[string]bool)
+	for _, role := range b {
+		remove[role.Name] = true
```

Comment:
Not being able to rename a role is a good argument if that should be supported by the operator in this way.
The name fallback is something I proposed elsewhere as well https://issues.redhat.com/browse/KEYCLOAK-16624 but in this case I think it would have confusing side effects as the ID in the resource (although respected if present) is always set to the resulting ID on creation https://github.com/keycloak/keycloak-operator/blob/2c058104137c861a1c64d721489d1bc3a1386cb1/pkg/common/cluster_actions.go#L133
Or am I confusing different kinds of IDs here?
I think the difference between renaming and recreating roles would be that the latter implicitly removes all associations with the role, correct?

Reply:
> I think the difference between renaming and recreating roles would be that the latter implicitly removes all associations with the role, correct?

Good point. I didn't realize this.
So I think we really need to rely on the IDs with the name as a fallback. The fallback could be used only in the `roleDifference` function. The name would never be set as the ID if that's what you meant that would be confusing.

The question is whether we really support roles renaming. The UI (Admin Console) doesn't support this but the REST API apparently does. So just to be on the safe side, I think operator should be ready for it.

Reply:
While I appreciate having the functionality to rename roles I think this makes things rather intransparent as matching with fallback has some ambiguities here. How for example should the operator even behave when reading
```
role(id1, name1)
```
from the server and getting
```
role(id1)
role(name1)
```
from the CR?
Furthermore, in the spirit of having easily readable CRs I would prefer being able to create and update roles with just a name and the non-default fields set, so I don't think forcing IDs would be the best solution either.
I'd like to keep things as they are now (matching by name), though this does of course sacrifice the (not available to the UI) possibility to rename roles while keeping all associations (which would get lost on re-create).
Any other opinions on this?

Reply:
I would still like to support role renaming, even though it might be a corner case. IMHO since it is supported at the API side so we should support it in the operator as well.

How about this.
* Use `role.Id` instead of `role.Name` in `roleDifference`.
* Some new roles already have ID set. And if some role in the CR doesn't have ID, take it as a new role.

WDYT?

Reply:
I'll rewrite the whole reconciliation function to make renames possible. It will have further implications like specifying a non-matching ID leading to re-creation. I will continue this later.

Reply:
This sounds good to me.

Reply:
> the ID in the resource (although respected if present)

correction: while writing the test it was not
This did not cause any new problems, just to clarify in case someone comes across this thread.